src/Microsoft.Diagnostics.EventFlow.Core/Implementations/LossReportingPropagatorBlock.cs

```diff
@@ -0,0 +1,209 @@
+        }
+
+        public Task Completion
+        {
+            get
+            {
+                if (this.propagateCompletion)
+                {
+                    Debug.Assert(this.target != null);
+                    return this.target.Completion;
+                }
+                else
+                {
+                    return this.completionSource.Value.Task;
+                }
+            }
+        }
+
+        public void Complete()
+        {
+            this.completed = true;
+
+            if (this.propagateCompletion)
+            {
+                Debug.Assert(this.target != null);
+                this.target.Complete();
+            }
+            else
+            {
+                this.completionSource.Value.TrySetResult(true);
+            }
+        }
+
+        public void Fault(Exception exception)
+        {
+            this.completed = true;
+
+            if (this.propagateCompletion)
+            {
+                Debug.Assert(this.target != null);
+                this.target.Fault(exception);
+            }
+            else
+            {
+                this.completionSource.Value.TrySetException(exception);
+            }
+        }
+
+        public IDisposable LinkTo(ITargetBlock<T> target, DataflowLinkOptions linkOptions)
+        {
+            Requires.NotNull(target, nameof(target));
+            if (this.target != null)
+            {
+                throw new InvalidOperationException($"{nameof(LossReportingPropagatorBlock<T>)} only supports a single target");
+            }
+
+            this.target = target;
+            if (linkOptions != null)
+            {
+                // Note DataflowLinkOptions.Append does not matter given that we are working with a single target only 
+                // and we will not enforce DataflowLinkOptions.MaxMessages
+                this.propagateCompletion = linkOptions.PropagateCompletion;
+            }
+
+            // We do not support re-linking
+            return new EmptyDisposable();
+        }
+
+        public DataflowMessageStatus OfferMessage(DataflowMessageHeader messageHeader, T messageValue, ISourceBlock<T> source, bool consumeToAccept)
+        {
+            VerifySourceInitiatedOperation(source);
+
+            if (this.completed)
+            {
+                return DataflowMessageStatus.DecliningPermanently;
+            }
+
+            var targetResponse = this.target.OfferMessage(messageHeader, messageValue, this, consumeToAccept);
+            switch(targetResponse)
+            {
+                case DataflowMessageStatus.Accepted:
+                    // Great, nothing to do
+                    break;
+
+                case DataflowMessageStatus.Declined:
+                case DataflowMessageStatus.DecliningPermanently:
+                case DataflowMessageStatus.NotAvailable:
+                    this.healthReporter.ReportThrottling();
```

Comment:
I feel like the switch is not needed here and just pass the call to the target. Whether the events will be dropped can determined by the return value BroadCastBlock's ConsumeMessage solely. To be more specific
1) NotAvailable will be a subsequent result of a failed ConsumeMessage call, so this will be duplicated with the ReportThrottling in ConsumeMessage.
2) It won't return DecliningPermanently, that happens when the target block is completed, when that happens, the current block is also completed.
3) It won't return Declined. Since there is always a source (unlike the first BufferBlock in the pipeline), it will return Postponed if the message is bounded by capacity.